File: Day10/day10.py
import sys


def preprocess_instructions(instructions):
    instructions = instructions.split("\n")
    instructions.remove("")
    instructions = [instruction.split(" ") for instruction in instructions]
    parsed_instructions = []
    for instruction in instructions:
        new_instruction = (
            [instruction[0], int(instruction[1])]
            if len(instruction) == 2
            else instruction
        )
        parsed_instructions.append(new_instruction)

    return parsed_instructions


def signal_strength_sum(instructions):
    registry_x = 1
    strength_sum = 0
    cycle_count = 0

    for instruction in instructions:
        cycle_count += 1
        if cycle_count % 40 == 20:
            strength_sum += cycle_count * registry_x
        if len(instruction) == 2:
            cycle_count += 1
            if cycle_count % 40 == 20:
                strength_sum += cycle_count * registry_x
            registry_x += instruction[1]
    return strength_sum


def move_sprite(registry):
    return [registry - 1, registry, registry + 1]


def line_and_pixel(cycle_count):
    return [cycle_count // 40, cycle_count % 40]


def pixel(is_on: bool):
    return "#" if is_on else "."


def check_new_line(cycle_count):
    return "\n" if cycle_count % 40 == 0 else ""


def render_display(instructions):
    display = ""
    registry_x = 1
    cycle_count = 0
    sprite_positions = [0, 1, 2]

    for instruction in instructions:
        display += check_new_line(cycle_count)
        display += pixel(cycle_count % 40 in sprite_positions)
        cycle_count += 1
        if len(instruction) == 2:
            display += check_new_line(cycle_count)
            display += pixel(cycle_count % 40 in sprite_positions)
            cycle_count += 1
            registry_x += instruction[1]
            sprite_positions = move_sprite(registry_x)
    return display


if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
    except IndexError:
        file_name = "input.txt"

    parsed_instructions = preprocess_instructions(open(file_name).read())

    strenth_sum = signal_strength_sum(parsed_instructions)
    print(f"The sum of signal strengths is {strenth_sum}.")

    print(f"\nHere's what the display looks like.")
    print(render_display(parsed_instructions))

File: Day10/day10.rs
fn get_data() -> String {
    let input_args: Vec<String> = std::env::args().collect();
    let mut file_name: String = "input.txt".to_string();
    if input_args.len() > 1 {
        file_name = input_args[1].clone();
    }
    let mut data = std::fs::read_to_string(file_name).unwrap();
    if data.ends_with("\n") {
        data.pop();
    }
    return data;
}

fn check_sprite(cycle_count: &i32, registry: &i32) -> bool {
    let position = cycle_count.rem_euclid(40);
    if (position == registry - 1) | (position == *registry) | (position == registry + 1) {
        return true
    }
    return false
}

fn pixel(is_on: bool) -> char {
    if is_on {
        return '#';
    } else {
        return  '.';
    }
}

fn check_new_line(cycle_count: &i32) -> bool {
    return cycle_count.rem_euclid(40) == 0
}

fn render_display(instructions: &Vec<&str>) -> (i32, String) {
    let mut display = String::new();
    let mut registry_x: i32 = 1;
    let mut strength_sum: i32 = 0;
    let mut cycle_count: i32 = 0;

    for instruction in instructions {
        let split_instruction: Vec<&str> = instruction.split(" ").collect();
        if check_new_line(&cycle_count) == true {
            display.push('\n');
        }
        display.push(pixel(check_sprite(&cycle_count, &registry_x)));
        cycle_count += 1;
        if cycle_count.rem_euclid(40) == 20 {
            strength_sum += cycle_count * registry_x
        }
        if split_instruction.len() == 2 {
            if check_new_line(&cycle_count) == true {
                display.push('\n');
            }
            display.push(pixel(check_sprite(&cycle_count, &registry_x)));
            cycle_count += 1;
            if cycle_count.rem_euclid(40) == 20 {
                strength_sum += cycle_count * registry_x
            }
            registry_x += split_instruction[1].parse::<i32>().unwrap();
        }
    }
    return (strength_sum, display)
}


fn main() {
    let data = get_data();
    let split_data: Vec<&str> = data.split("\n").collect();

    let (strength_sum, display) = render_display(&split_data);
    println!("The signal strength is {}", strength_sum);
    println!("\nThis is what the display looks like!");
    println!("{}", display)
}